File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */

    constructor(){
        this.lca = null
    }

    lowestCommonAncestor(root, p, q) {
        this.lca = root; 
        this.findLCA(root.left, p, q) ? this.lca = root.left : null;
        this.findLCA(root.right, p, q) ? this.lca = root.right : null;
        return this.lca;
    }

    findLCA(root, p, q){
        console.log(root?.val)
        return this.find(root, p) && this.find(root, q)
    }

    find(root, node){
        if(!root) return false; 
        if(root.val == node.val) return true; 
        let val = this.find(root.left, node)
        if(val) return true; 
        let val2 = this.find(root.right, node)
        if(val2) return true; 
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        is_all_balanced = True

        def is_balanced(node):
            nonlocal is_all_balanced
            if not node:
                return 0
            print(node.val)
            left_val = is_balanced(node.left)
            right_val = is_balanced(node.right)
            print(node.val, left_val, right_val)
            diff = right_val - left_val
            if abs(diff) > 1:
                is_all_balanced = False
            
            return 1 + max(left_val, right_val)
        is_balanced(root)

        return is_all_balanced
        